src/pymor/models/iosys.py

```diff
@@ -940,6 +940,483 @@ def get_ast_spectrum(self, ast_pole_data=None, mu=None):
+        error_estimator
+            An error estimator for the problem. This can be any object with an
+            `estimate_error(U, mu, model)` method. If `error_estimator` is not `None`, an
+            `estimate_error(U, mu)` method is added to the model which will call
+            `error_estimator.estimate_error(U, mu, self)`.
+        visualizer
+            A visualizer for the problem. This can be any object with a `visualize(U, model, ...)`
+            method. If `visualizer` is not `None`, a `visualize(U, *args, **kwargs)` method is added
+            to the model which forwards its arguments to the visualizer's `visualize` method.
+        name
+            Name of the system.
+
+        Returns
+        -------
+        phlti
+            The |PHLTIModel| with operators J, R, G, P, S, N, and E.
+        """
+        assert isinstance(J, (np.ndarray, sps.spmatrix))
+        assert isinstance(R, (np.ndarray, sps.spmatrix))
+        assert isinstance(G, (np.ndarray, sps.spmatrix))
+        assert isinstance(P, (np.ndarray, sps.spmatrix))
+        assert S is None or isinstance(S, (np.ndarray, sps.spmatrix))
+        assert N is None or isinstance(N, (np.ndarray, sps.spmatrix))
+        assert E is None or isinstance(E, (np.ndarray, sps.spmatrix))
+
+        J = NumpyMatrixOperator(J, source_id=state_id, range_id=state_id)
+        R = NumpyMatrixOperator(R, source_id=state_id, range_id=state_id)
+        G = NumpyMatrixOperator(G, range_id=state_id)
+        P = NumpyMatrixOperator(P, range_id=state_id)
+        if S is not None:
+            S = NumpyMatrixOperator(S)
+        if N is not None:
+            N = NumpyMatrixOperator(N)
+        if E is not None:
+            E = NumpyMatrixOperator(E, source_id=state_id, range_id=state_id)
+
+        return cls(J, R, G, P, S, N, E,
+                   solver_options=solver_options, error_estimator=error_estimator, visualizer=visualizer,
+                   name=name)
+
+    def to_matrices(self):
+        """Return operators as matrices.
+
+        Returns
+        -------
+        J
+            The |NumPy array| or |SciPy spmatrix| J.
+        R
+            The |NumPy array| or |SciPy spmatrix| R.
+        G
+            The |NumPy array| or |SciPy spmatrix| G.
+        P
+            The |NumPy array| or |SciPy spmatrix| P.
+        S
+            The |NumPy array| or |SciPy spmatrix| S or `None` (if Cv is a `ZeroOperator`).
+        N
+            The |NumPy array| or |SciPy spmatrix| N or `None` (if Cv is a `ZeroOperator`).
+        E
+            The |NumPy array| or |SciPy spmatrix| E.
+        """
+        J = to_matrix(self.J)
+        R = to_matrix(self.R)
+        G = to_matrix(self.G)
+        P = to_matrix(self.P)
+        S = None if isinstance(self.S, ZeroOperator) else to_matrix(self.S)
+        N = None if isinstance(self.N, ZeroOperator) else to_matrix(self.N)
+        E = to_matrix(self.E)
+
+        return J, R, G, P, S, N, E
+
+    @cached
+    def to_lti(self):
```

Comment:
It think we need to be a little bit more careful with the caching:

Caching an object with cached methods has never really been envisioned and is somewhat problematic. Unless a persistent id is given for the cached object, the keys of for its cached methods will depend on its `uid`. `MemoryRegion` uses `deepcopy` to store and retrieve values, which should result in the same `uid`, I believe. (If it should is another question.) `DiskRegion`, however, pickles, which will result in a new `uid` so the cached values for the `LTIModel` will be lost.

The second problem is that by caching both `to_lti` and `gramian`, etc. in `PHLTIModel`, the results will be stored twice in memory.

Overall, I would propose to

- Cache `to_lti` by just creating `_lti_model` in `__init__` and returning it in `to_lti`.
- Remove `@cached` from the methods in `PHLTIModel`.

Reply:
> MemoryRegion uses deepcopy to store and retrieve values, which should result in the same uid, I believe. (If it should is another question.)

I forgot: `__deepcopy__` always returns `self` for `ImmutableObjects`, so that case should be fine. The problem with `DiskRegion` remains.

Reply:
Okay sounds reasonable. However, I have oriented myself to the `SecondOrderModel`. There it should be the same.

Reply:
I'am fine with your proposals. Should I implement them?

Reply:
@pmli, what do you think? We can also leave it as is, and then try to handle both cases in later pr.

Reply:
@sdrave Your proposal sounds good to me (I didn't realize the issue in `SecondOrderModel` before). Doing the same for `SecondOrderModel` would then also be good, except it might be ok for the `gramian` method there to be cached since it does something besides calling `to_lti`. I think it is fine to do it in this PR as some changes were already done to `SecondOrderModel`.